id: python.lang.correctness.cannot-cache-generators.cannot-cache-generators
patterns:
- pattern-inside: "@functools.lru_cache(...)\ndef $FUNC(...):\n    ...\n    yield\
    \ ...\n"
- pattern: functools.lru_cache(...)
message: Generators can only be consumed once, so in most cases, caching them will
  cause an error when the already-consumed generator is retrieved from cache.
languages:
- python
severity: WARNING
metadata:
  category: correctness
  technology:
  - python
  license: Commons Clause License Condition v1.0[LGPL-2.1-only]
  source: https://semgrep.dev/r/python.lang.correctness.cannot-cache-generators.cannot-cache-generators
  shortlink: https://sg.run/gG8y
  semgrep.dev:
    rule:
      r_id: 20041
      rv_id: 110327
      rule_id: 4bUe9x
      version_id: ExTjNBX
      url: https://semgrep.dev/playground/r/ExTjNBX/python.lang.correctness.cannot-cache-generators.cannot-cache-generators
      origin: community
